### PR TITLE
Add Texture support to node explorer

### PR DESCRIPTION
**Pull request by levi**
This is a first pass at adding richer inspection of apps using Texture view/layer backed nodes in their hierarchy. It generalizes the View type with the introduction of the `FLEXElement` proxy class. Items represent either `UIView`s or `ASDisplayNode`s and provide a public interface that encapsulates the heuristics of jumping in and out of the node and view hierarchies. Goal was to add support without having to include the AsyncDisplayKit framework as a dependency, hence the excessive amount of `performSelector:` and `NSInvocation`.

I'd like to add some richer properties to the custom node object explorer view controller, but I wanted to put this up in its simplest form to get feedback on the direction and start the conversation about including node hierarchy support to FLEX. Tons of great ideas came to mind while working on this that I'd love to see introduced into FLEX — opening up the object explorer view controller API to be extensible by consumers of the framework being one of them. Looking forward to making it easier to debug Texture apps with FLEX!

- [ ] elementsAtTapPoint equality always fails. Compare element objects.

### CODE REVIEW

**Comment by appleguy** [comment]
@levi very cool!  This could be useful to thousands of developers :).  If it gets merged, let's feature it on asyncdisplaykit.org and in the Slack channel.

Could you share some screenshots of what it looks like for an end user?

**Comment by ryanolsonk** [comment]
Super cool! I'm on holiday right now but will try to get this reviewed asap! It would be awesome to see screenshots as Scott mentioned.

**Comment by ryanolsonk** [comment]
I would also love to gut the current object explorer architecture as it is painful to work with or extend as-is. I've got a branch rebuilding the explorer VCs with IGListKit but I haven't had time to finish it. Hopefully would be easier to customize the node explorer in that new-world. Please don't judge me for the code I wrote three years ago 😉.

One thought from reading the description is whether we could make `FLEXHierarchyItem` a protocol that anything (view, node, component, etc.) could conform to. We could possibly ship the implementation for the `ASNode` conformance in a category inside of ASDK. That might cut out some of the runtime introspection. Just an idea so let me know what you think.

**Comment by levi** [comment]
@ryanolsonk I love the idea of extending the `FLEXHierarchyItem` to be a protocol, such that any framework can support hierarchy inspection. There is a caveat, however, where I've extended FLEX to jump between view and node hierarchies based on some heuristic rules for the entire picture to be visible in the explorer. If FLEX were to be opened up to support multiple frameworks, we'd have to figure out how to support custom traces of the hierarchy in a meaningful way — likely just push that responsibility down to the individual framework implementing support.

I haven't had much time to look at this since I've put it up, but happy to continue the dialog with you to figure out where to take this. In the meantime, we've started using my branch internally at Pinterest and it's working great. Excited to provide the support to others in the near future.

**Comment by ryanolsonk** [comment]
Finally had a chance to patch this locally and did some quick checks on the example project to look for functional regressions. Here are a few bugs I noticed:

Info bar doesn't show up below toolbar when an item is selected:
screenshot 2017-01-15 16 34 34
vs.
screenshot 2017-01-15 16 36 01

When an item is selected, the hierarchy list lacks the depth and selection indicators
screenshot 2017-01-15 16 34 46
vs
screenshot 2017-01-15 16 36 08

When the hierarchy list is dismissed, outlines of the other views at the tap point disappear
screenshot 2017-01-15 16 34 52

When moving a selected item, the outline/selection doesn't move along with the item
screenshot 2017-01-15 16 35 07
vs
screenshot 2017-01-15 16 36 26


I'll try to track down fixes for these if I get some time. Looking forward to getting this out to the community!

File: Classes/ObjectExplorers/FLEXObjectExplorerFactory.m [hunk 1]
```diff
@@ -31,12 +32,14 @@ + (FLEXObjectExplorerViewController *)explorerViewControllerForObject:(id)object
...
     static dispatch_once_t once;
     dispatch_once(&once, ^{
+        Class displayNodeClass = NSClassFromString(@"ASDisplayNode");
         explorerSubclassesForObjectTypeStrings = @{NSStringFromClass([NSArray class])          : [FLEXArrayExplorerViewController class],
                                                    NSStringFromClass([NSSet class])            : [FLEXSetExplorerViewController class],
                                                    NSStringFromClass([NSDictionary class])     : [FLEXDictionaryExplorerViewController class],
                                                    NSStringFromClass([NSUserDefaults class])   : [FLEXDefaultsExplorerViewController class],
                                                    NSStringFromClass([UIViewController class]) : [FLEXViewControllerExplorerViewController class],
                                                    NSStringFromClass([UIView class])           : [FLEXViewExplorerViewController class],
+                                                   NSStringFromClass(displayNodeClass)         : [FLEXNodeExplorerViewController class],
```
Comment (ryanolsonk) [inline Classes/ObjectExplorers/FLEXObjectExplorerFactory.m]: This crashes for projects that don't import ASDK due to the nil key.

File: Classes/Toolbar/FLEXExplorerToolbar.m [hunk 2]
```diff
@@ -147,28 +147,28 @@ - (void)layoutSubviews
...
-    self.selectedViewDescriptionLabel.frame = descriptionLabelFrame;
+    descriptionLabelFrame.size.width = CGRectGetMaxX(self.selectedItemDescriptionContainer.bounds) - kHorizontalPadding - descriptionOriginX;
+    self.selectedItemDescriptionLabel.frame = descriptionLabelFrame;
 }
 
 
 #pragma mark - Setter Overrides
 
-- (void)setSelectedViewOverlayColor:(UIColor *)selectedViewOverlayColor
+- (void)setselectedItemOverlayColor:(UIColor *)selectedItemOverlayColor
```
Comment (ryanolsonk) [inline Classes/Toolbar/FLEXExplorerToolbar.m]: Here's the source of the toolbar info not showing up. Should be `setSelectedItemOverlayColor` (capitalization)

File: Classes/Toolbar/FLEXExplorerToolbar.m [hunk 3]
```diff
@@ -147,28 +147,28 @@ - (void)layoutSubviews
...
-        _selectedViewOverlayColor = selectedViewOverlayColor;
-        self.selectedViewColorIndicator.backgroundColor = selectedViewOverlayColor;
+    if (![_selectedItemOverlayColor isEqual:selectedItemOverlayColor]) {
+        _selectedItemOverlayColor = selectedItemOverlayColor;
+        self.selectedViewColorIndicator.backgroundColor = selectedItemOverlayColor;
     }
 }
 
-- (void)setSelectedViewDescription:(NSString *)selectedViewDescription
+- (void)setselectedItemDescription:(NSString *)selectedItemDescription
```
Comment (ryanolsonk) [inline Classes/Toolbar/FLEXExplorerToolbar.m]: Same thing here

File: Classes/ExplorerInterface/FLEXExplorerViewController.m [hunk 4]
```diff
@@ -329,27 +329,27 @@ - (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(N
 
 - (void)updateOverlayAndDescriptionForObjectIfNeeded:(id)object
 {
-    NSUInteger indexOfView = [self.viewsAtTapPoint indexOfObject:object];
-    if (indexOfView != NSNotFound) {
-        UIView *view = self.viewsAtTapPoint[indexOfView];
-        NSValue *key = [NSValue valueWithNonretainedObject:view];
-        UIView *outline = self.outlineViewsForVisibleViews[key];
+    NSUInteger indexOfItem = [self.itemsAtTapPoint indexOfObject:object];
```
Comment (ryanolsonk) [inline Classes/ExplorerInterface/FLEXExplorerViewController.m]: This lookup fails because it's looking for a view (`object`) in the items array. One approach would be to pass the item in the `context` parameter of the KVO.

**Comment by levi** [comment]
Thanks for the review, @ryanolsonk. I'll take a look at fixing this up in the next week and will get back to you.

**Comment by ryanolsonk** [comment]
Great, thanks @levi!

File: Classes/ObjectExplorers/FLEXObjectExplorerFactory.m (hunk 1 above)
Comment (levi) [inline Classes/ObjectExplorers/FLEXObjectExplorerFactory.m]: Fixed

File: Classes/Toolbar/FLEXExplorerToolbar.m (hunk 2 above)
Comment (levi) [inline Classes/Toolbar/FLEXExplorerToolbar.m]: Fixed

File: Classes/Toolbar/FLEXExplorerToolbar.m (hunk 3 above)
Comment (levi) [inline Classes/Toolbar/FLEXExplorerToolbar.m]: Fixed

File: Classes/ExplorerInterface/FLEXExplorerViewController.m (hunk 4 above)
Comment (levi) [inline Classes/ExplorerInterface/FLEXExplorerViewController.m]: Fixed

**Comment by levi** [comment]
@ryanolsonk finally found some time to fix up the issues in this diff. Going to do a test pass on my side this week to identify remaining issues. Is this something you'd like to land?

**Comment by ryanolsonk** [comment]
Awesome! Will patch and test out locally. I'm keen to do a full review pass and land this once we've verified all the functionality is working!

**Comment by levi** [comment]
That's great news! There are still a few edge cases I need to work out. Can prioritize fixing them up this week for a full review.

**Comment by NSExceptional** [comment]
Bump, @levi?

**Comment by NSExceptional** [comment]
@larrytin It looks like you had this almost done. Is there anything I can do to help get this merged?

**Comment by NSExceptional** [comment]
@levi It looks like you had this almost done. Is there anything I can do to help get this merged?

**Comment by levi** [comment]
It’s been a minute since I looked at this thing. I don’t have the time to see what the state of it is and if it still works with Texture. Are you able to take it on?

**Comment by NSExceptional** [comment]
I don't think so, no. I don't know the first thing about AsyncDisplayKit, etc, and I imagine this would be a lot of work for something pretty niche 😅

**Comment by ryanolsonk** [comment]
Going to close this out since we don't have plans to take it on at this time.
